### PR TITLE
create createCertificate.js

### PR DESCRIPTION
**Pull request by FabianKLS**
Helpful for creating x509 certificates for individual enrollments of device in device provisioning service. Is inline with example given in the device part of the provisioning documentation.

 

# Checklist
- [x] I have read the [contribution guidelines] (https://github.com/Azure/azure-iot-sdk-node/blob/master/.github/CONTRIBUTING.md).
- [x] I added or modified the existing tests to cover the change (we do not allow our test coverage to go down).
- If this is a modification that impacts the behavior of a public API
  - [ ] I edited the corresponding document in the `devdoc` folder and added or modified requirements.

# Reference/Link to the issue solved with this PR (if any)

# Description of the problem

# Description of the solution

### CODE REVIEW

File: provisioning/service/samples/createCertificate.js [hunk 1]
```diff
@@ -0,0 +1,31 @@
+'use strict';
```
Comment (pierreca) [inline provisioning/service/samples/createCertificate.js]: Please add the standard Microsoft copyright/license header is missing:
// Copyright (c) Microsoft. All rights reserved.
// Licensed under the MIT license. See LICENSE file in the project root for full license information.

**Comment by pierreca** [comment]
little nitpicky for sure but our file names are usually not camel-cased and instead snake_cased.

File: provisioning/service/samples/createCertificate.js [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+var pem = require('pem');
+var fs = require('fs');
+var certPath = __dirname + "\\cert\\";
```
Comment (pierreca) [inline provisioning/service/samples/createCertificate.js]: >__dirname + "\\cert\\"; [](start = 15, length = 23)

I would encourage using path.join or path.resolve instead to automatically deal with cross-platform issues.

**Comment by pierreca** [comment]
Thanks a lot for submitting this pull-request - it'll help users for sure. 
can you also update the readme with instructions on how/why use it?

A couple additional things to consider:
- since this is turning into sort of a command line utility, I would encourage you to look at yargs, and as an example what we've done in build/tools/create_device_certs.js - that makes it easy to create a CLI and "help" in case someone is "using it wrong"
- I'm not sure if this is meant to be a sample or a helper tool - it's fine in the samples folder for now but would be glad to hear if we need sort-of a "tools" folder.

File: provisioning/service/samples/createCertificate.js (hunk 1 above)
Comment (FabianKLS) [reply provisioning/service/samples/createCertificate.js]: Added

**Comment by FabianKLS** [comment]
I've made the changes your requested. Thanks for the feedback. The tool is a helper tool for the provisioning service. I wasn't aware of the build folder and the tools there. What is build/tools/create_device_certs.js exactly doing? There seems to be overlap.

**Comment by pierreca** [comment]
The `create_device_certs.js` script creates self-signed certificates (without setting the common name to the device id though) and adds the certificate thumbprint to the registry. it is used for end-to-end tests.

would you mind renaming your file to `create_certificates.js` (snake_case instead of camelCase) too? with that I think that'd be all and we could pull that change in.

File: provisioning/service/samples/createCertificate.js [hunk 3]
```diff
@@ -0,0 +1,33 @@
+// Copyright (c) Microsoft. All rights reserved.
```
Comment (BertKleewein) [inline provisioning/service/samples/createCertificate.js]: >Copyright [](start = 3, length = 9)

Rats!  I didn't see this PR until just now after I submitted a PR that included a similar tool.  See [https://github.com/Azure/azure-iot-sdk-node/pull/198](https://github.com/Azure/azure-iot-sdk-node/pull/198)

Comment (pierreca) [reply provisioning/service/samples/createCertificate.js]: yeah at that point we probably need to merge these tools.

And then consider merging this into iothub-explorer.

**Comment by FabianKLS** [comment]
I've renamed the file.

File: provisioning/service/samples/createCertificate.js (hunk 3 above)
Comment (BertKleewein) [reply provisioning/service/samples/createCertificate.js]: Disagree on putting this into iothub-explorer.  This is just a wrapper around PEM and has no actual connection to the hub code.  Also, these certs are test-only.  We don't want to get into the business of telling people how to make production certs.  They should have their own process for that.  

---
In reply to: [161137256](https://github.com/Azure/azure-iot-sdk-node/pull/196#discussion_r161137256) [](ancestors = 161137256)

Comment (pierreca) [reply provisioning/service/samples/createCertificate.js]: fair enough - thanks for chiming in :) 

---
In reply to: [161297082](https://github.com/Azure/azure-iot-sdk-node/pull/196#discussion_r161297082) [](ancestors = 161297082,161137256)

**Comment by pierreca** [comment]
@FabianKLS would you mind reviewing #198 and see if it meets the need that you were fulfilling with this PR too? we're going to need other parts of #198 so we're going to pull it anyway, the question is does it make sense to have your script too?

**Comment by FabianKLS** [comment]
I've used the tool and it's working for me. So this request can be closed/removed.
Only 'request' for #198 is the ability to provide a folder where the certificates are stored.  After 198 is published I will see if I can add this. Oke?

**Comment by pierreca** [comment]
Thanks @FabianKLS. #198 has been merged. whenever you want to add the folder feature, just go for it and @BertKleewein will review it. Closing this one.

**Pull request closed without merging**